CSC8503CoreClasses/RecastBuilder.cpp

```diff
@@ -0,0 +1,301 @@
+#include "RecastBuilder.h"
+#include "RenderObject.h"
+#include "GameObject.h"
+#include "../OpenGLRendering/OGLRenderer.h"
+#include "../Detour/Include/DetourNavMeshBuilder.h"
+#include "../Detour/Include/DetourNavMeshQuery.h"
+
+using namespace NCL::CSC8503;
+
+RecastBuilder::RecastBuilder() {
+	mTriAreas = nullptr;
+	mSolid = nullptr;
+	mCompHF = nullptr;
+	mContSet = nullptr;
+	mPolyMesh = nullptr;
+	mMeshDetail = nullptr;
+	mNavMesh = nullptr;
+	mNavMeshQuery = nullptr;
+}
+
+RecastBuilder::~RecastBuilder() {
+	cleanup();
+}
+
+void RecastBuilder::BuildNavMesh(std::vector<GameObject*> objects) {
+	if (objects.empty()) return;
+
+	cleanup();
+
+	// Translating the meshes into useable values for Recast
+	float* bmin = new float[3] {std::numeric_limits<float>::max(), std::numeric_limits<float>::max(), std::numeric_limits<float>::max()};
+	float* bmax = new float[3] {-std::numeric_limits<float>::max(), -std::numeric_limits<float>::max(), -std::numeric_limits<float>::max()};
+	int currentTriCount = 0;
+	int vertCount = 0;
+	int trisCount = 0;
+	for (int i = 0; i < objects.size(); i++) {
+		vertCount += objects[i]->GetRenderObject()->GetMesh()->GetVertexCount();
+		trisCount += objects[i]->GetRenderObject()->GetMesh()->GetPrimitiveCount();
+	}
+	float* verts = new float[vertCount * 3];
+	unsigned int* tris = new unsigned int[trisCount * 3];
+	int vCount = 0;
+	int tCount = 0;
+	vertCount = 0;
+	for (int i = 0; i < objects.size(); i++) {
+		int lastVertCount = vertCount;
+		std::vector<Vector3> vertsVector = objects[i]->GetRenderObject()->GetMesh()->GetPositionData();
+		for (int j = 0; j < vertsVector.size(); j++) {
+			verts[vCount++] = (vertsVector[j].x * objects[i]->GetTransform().GetScale().x) + objects[i]->GetTransform().GetPosition().x;
+			verts[vCount++] = (vertsVector[j].y * objects[i]->GetTransform().GetScale().y) + objects[i]->GetTransform().GetPosition().y;
+			verts[vCount++] = (vertsVector[j].z * objects[i]->GetTransform().GetScale().z) + objects[i]->GetTransform().GetPosition().z;
+			bmin[0] = std::min(bmin[0], verts[vCount - 3]);
+			bmin[1] = std::min(bmin[1], verts[vCount - 2]);
+			bmin[2] = std::min(bmin[2], verts[vCount - 1]);
+			bmax[0] = std::max(bmax[0], verts[vCount - 3]);
+			bmax[1] = std::max(bmax[1], verts[vCount - 2]);
+			bmax[2] = std::max(bmax[2], verts[vCount - 1]);
```

Comment:
We can add constant variables to the anon namespace for the magic numbers(indexes) for clearing out what they mean 🏎️